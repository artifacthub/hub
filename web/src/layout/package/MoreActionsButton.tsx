import classnames from 'classnames';
import React, { useEffect, useRef, useState } from 'react';
import { BiCode } from 'react-icons/bi';
import { HiDotsVertical } from 'react-icons/hi';

import useOutsideClick from '../../hooks/useOutsideClick';
import { SearchFiltersURL } from '../../types';
import styles from './MoreActionsButton.module.css';
import WidgetModal from './WidgetModal';

interface Props {
  packageId: string;
  packageName: string;
  packageDescription: string;
  visibleWidget: boolean;
  searchUrlReferer?: SearchFiltersURL;
  fromStarredPage?: boolean;
}

const MoreActionsButton = (props: Props) => {
  const [openStatus, setOpenStatus] = useState(false);
  const [visibleWidget, setVisibleWidget] = useState<boolean>(props.visibleWidget);
  const [currentPkgId, setCurrentPkgId] = useState<string>(props.packageId);

  const ref = useRef(null);
  useOutsideClick([ref], openStatus, () => setOpenStatus(false));

  useEffect(() => {
    if (props.packageId !== currentPkgId && openStatus) {
      setVisibleWidget(false);
      setCurrentPkgId(props.packageId);
    }
  }, [props.packageId]); /* eslint-disable-line react-hooks/exhaustive-deps */

  return (
    <>
      <div className="d-none d-md-block position-relative ml-2">
        <button
          data-testid="moreActionsBtn"
          className="btn p-0 position-relative"
          type="button"
          onClick={() => {
            setOpenStatus(true);
          }}
        >
          <div className="d-flex flex-row align-items-center justify-content-center">
            <div
              className={`rounded-circle d-flex align-items-center justify-content-center text-primary iconSubsWrapper ${styles.iconWrapper}`}
            >
              <HiDotsVertical />
            </div>
          </div>
        </button>

        <div
          ref={ref}
          data-testid="subsBtnDropdown"
          className={classnames('dropdown-menu dropdown-menu-right p-0', styles.dropdown, { show: openStatus })}
        >
          <div className={`arrow ${styles.arrow}`} />

          <button
            className="dropdown-item btn btn-sm rounded-0 text-secondary"
            onClick={() => {
              setVisibleWidget(true);
              setOpenStatus(false);
            }}
          >
            <div className="d-flex flex-row align-items-center">
              <BiCode className="mr-2" />
              <div>Embed widget</div>
            </div>
          </button>
        </div>
      </div>

      <WidgetModal {...props} visibleWidget={visibleWidget} setOpenStatus={setVisibleWidget} />
    </>
  );
};

export default MoreActionsButton;
