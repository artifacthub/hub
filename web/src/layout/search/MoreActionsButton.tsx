import classnames from 'classnames';
import React, { useRef, useState } from 'react';
import { BiCode } from 'react-icons/bi';
import { HiDotsVertical } from 'react-icons/hi';

import useOutsideClick from '../../hooks/useOutsideClick';
import styles from './MoreActionsButton.module.css';
import WidgetsGroupModal from './WidgetsGroupModal';

const MoreActionsButton = () => {
  const [openStatus, setOpenStatus] = useState(false);
  const [visibleWidget, setVisibleWidget] = useState<boolean>(false);

  const ref = useRef(null);
  useOutsideClick([ref], openStatus, () => setOpenStatus(false));

  return (
    <>
      <div className="d-none d-md-block position-relative ml-2">
        <button
          data-testid="moreActionsBtn"
          className="btn p-0 position-relative"
          type="button"
          onClick={() => {
            setOpenStatus(true);
          }}
        >
          <div className="d-flex flex-row align-items-center justify-content-center">
            <div
              className={`rounded-circle d-flex align-items-center justify-content-center text-primary iconSubsWrapper ${styles.iconWrapper}`}
            >
              <HiDotsVertical />
            </div>
          </div>
        </button>

        <div
          ref={ref}
          data-testid="subsBtnDropdown"
          className={classnames('dropdown-menu dropdown-menu-right p-0', styles.dropdown, { show: openStatus })}
        >
          <div className={`arrow ${styles.arrow}`} />

          <button
            className="dropdown-item btn btn-sm rounded-0 text-secondary"
            onClick={() => {
              setVisibleWidget(true);
              setOpenStatus(false);
            }}
          >
            <div className="d-flex flex-row align-items-center">
              <BiCode className="mr-2" />
              <div>Embed results</div>
            </div>
          </button>
        </div>
      </div>

      <WidgetsGroupModal visibleWidget={visibleWidget} setOpenStatus={setVisibleWidget} />
    </>
  );
};

export default MoreActionsButton;
