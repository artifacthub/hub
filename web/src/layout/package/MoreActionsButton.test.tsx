import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';

import MoreActionsButton from './MoreActionsButton';

const defaultProps = {
  packageId: 'id',
  packageName: 'pkg',
  packageDescription: 'this is the description',
  visibleWidget: false,
};

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useNavigate: () => mockUseNavigate,
}));

const openMock = jest.fn();
window.open = openMock;

describe('MoreActionsButton', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <MoreActionsButton {...defaultProps} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <MoreActionsButton {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: 'Open menu' });
      expect(btn).toBeInTheDocument();
    });

    it('displays dropdown', async () => {
      render(
        <Router>
          <MoreActionsButton {...defaultProps} />
        </Router>
      );

      const dropdown = screen.getByRole('menu');
      expect(dropdown).toBeInTheDocument();
      expect(dropdown).not.toHaveClass('show');

      const btn = screen.getByRole('button', { name: 'Open menu' });
      expect(btn).toBeInTheDocument();

      await userEvent.click(btn);

      expect(dropdown).toHaveClass('show');
      expect(screen.getByText('Embed widget')).toBeInTheDocument();
    });

    it('opens modal', async () => {
      render(
        <Router>
          <MoreActionsButton {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: 'Open menu' });
      expect(btn).toBeInTheDocument();

      await userEvent.click(btn);

      const widgetBtn = screen.getByRole('button', { name: 'Open embed widget modal' });
      expect(widgetBtn).toBeInTheDocument();

      await userEvent.click(widgetBtn);

      expect(screen.getByRole('dialog')).toBeInTheDocument();

      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith('?modal=widget', { replace: true, state: null });
    });

    describe('report abuse', () => {
      it('opens url', async () => {
        Object.defineProperty(document, 'querySelector', {
          value: (selector: any) => {
            switch (selector) {
              case `meta[name='artifacthub:reportURL']`:
                return {
                  getAttribute: () => 'http://test.com',
                };
              default:
                return false;
            }
          },
          writable: true,
        });

        render(
          <Router>
            <MoreActionsButton {...defaultProps} />
          </Router>
        );

        const dropdown = screen.getByRole('menu');
        expect(dropdown).toBeInTheDocument();
        expect(dropdown).not.toHaveClass('show');

        const btn = screen.getByRole('button', { name: 'Open menu' });
        expect(btn).toBeInTheDocument();

        await userEvent.click(btn);

        expect(dropdown).toHaveClass('show');

        const reportURLBtn = screen.getByRole('button', { name: 'Open report abuse url' });
        expect(reportURLBtn).toBeInTheDocument();

        await userEvent.click(reportURLBtn);

        expect(openMock).toHaveBeenCalledTimes(1);
        expect(openMock).toHaveBeenCalledWith('http://test.com', '_blank');
      });

      it('does not render when url is undefined', async () => {
        Object.defineProperty(document, 'querySelector', {
          value: (selector: any) => {
            switch (selector) {
              case `meta[name='artifacthub:reportURL']`:
                return {
                  getAttribute: () => null,
                };
              default:
                return false;
            }
          },
          writable: true,
        });

        render(
          <Router>
            <MoreActionsButton {...defaultProps} />
          </Router>
        );

        const dropdown = screen.getByRole('menu');
        expect(dropdown).toBeInTheDocument();
        expect(dropdown).not.toHaveClass('show');

        const btn = screen.getByRole('button', { name: 'Open menu' });
        expect(btn).toBeInTheDocument();

        await userEvent.click(btn);

        expect(dropdown).toHaveClass('show');
        expect(screen.queryByRole('button', { name: 'Open report abuse url' })).toBeNull();
      });
    });

    it('does not render when url is an empty string', async () => {
      Object.defineProperty(document, 'querySelector', {
        value: (selector: any) => {
          switch (selector) {
            case `meta[name='artifacthub:reportURL']`:
              return {
                getAttribute: () => '',
              };
            default:
              return false;
          }
        },
        writable: true,
      });

      render(
        <Router>
          <MoreActionsButton {...defaultProps} />
        </Router>
      );

      const dropdown = screen.getByRole('menu');
      expect(dropdown).toBeInTheDocument();
      expect(dropdown).not.toHaveClass('show');

      const btn = screen.getByRole('button', { name: 'Open menu' });
      expect(btn).toBeInTheDocument();

      await userEvent.click(btn);

      expect(dropdown).toHaveClass('show');
      expect(screen.queryByRole('button', { name: 'Open report abuse url' })).toBeNull();
    });
  });
});
