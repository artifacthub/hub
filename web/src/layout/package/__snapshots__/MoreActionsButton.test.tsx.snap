// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreActionsButton creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-none d-lg-block position-relative ms-2"
  >
    <button
      aria-expanded="false"
      aria-label="Open menu"
      class="btn btn-outline-primary p-0 position-relative lh-1 fs-5 iconWrapper"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z"
        />
      </svg>
    </button>
    <div
      class="dropdown-menu dropdown-menu-end p-0 dropdown"
      role="menu"
    >
      <div
        class="dropdown-arrow arrow"
      />
      <button
        aria-label="Open embed widget modal"
        class="dropdown-item btn btn-sm rounded-0 text-dark"
      >
        <div
          class="d-flex flex-row align-items-center"
        >
          <svg
            class="me-2 position-relative icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.293 6.293 2.586 12l5.707 5.707 1.414-1.414L5.414 12l4.293-4.293zm7.414 11.414L21.414 12l-5.707-5.707-1.414 1.414L18.586 12l-4.293 4.293z"
            />
          </svg>
          <div>
            Embed widget
          </div>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
